RefApp/ViewControllers/AssetListTableViewController.swift

```diff
@@ -0,0 +1,171 @@
+    ///   - query: The optional query to filter by Ex:assetType=TV_CHANNEL
+    ///   - environment: Customer specific *Exposure* environment
+    ///   - endpoint: Base exposure url. This is the customer specific URL to Exposure
+    ///   - method: http method - GET
+    fileprivate func loadAssets(query: String, environment: Environment, endpoint: String, method: HTTPMethod) {
+        ExposureApi<AssetList>(environment: environment,
+                               endpoint: endpoint,
+                               query: query,
+                               method: method)
+            .request()
+            .validate()
+            .response { [weak self] in
+                
+                if let value = $0.value {
+                    self?.assets = value.items ?? []
+                    self?.assetsDidLoad(value.items ?? [])
+                }
+                
+                if let error = $0.error {
+                    let okAction = UIAlertAction(title: NSLocalizedString("Ok", comment: ""), style: .cancel, handler: {
+                        (alert: UIAlertAction!) -> Void in
+                        self?.assetsDidLoad([])
+                    })
+                    
+                    let message = "\(error.code) " + error.message + "\n" + (error.info ?? "")
+                    self?.popupAlert(title: error.domain , message: message, actions: [okAction], preferedStyle: .alert)
+                }
+        }
+    }
+    
+    
+    /// Create the datasource for the tableview
+    ///
+    /// - Parameter assets: loaded assets from Exposure API
+    func assetsDidLoad(_ assets: [Asset]) {
+        datasource = .make(for: assets)
+        DispatchQueue.main.async {
+            self.tableView.dataSource = self.datasource
+            self.tableView.reloadData()
+        }
+    }
+}
+
+// MARK: - Actions
+extension AssetListTableViewController {
+    
+    /// User confirmation for logout
+    @objc fileprivate func handleLogout() {
+        let title = NSLocalizedString("Log out", comment: "")
+        let message = NSLocalizedString("Do you want to log out from the application ?", comment: "")
+        
+        let logOutAction = UIAlertAction(title: NSLocalizedString("Yes", comment: ""), style: UIAlertAction.Style.default, handler: { alert -> Void in
+            self.logoutUser()
+        })
+        
+        let cancelAction = UIAlertAction(title: NSLocalizedString("No", comment: ""), style: UIAlertAction.Style.default, handler: {
+            (action : UIAlertAction!) -> Void in })
+        
+        self.popupAlert(title: title, message: message, actions: [logOutAction, cancelAction])
+    }
+    
+    /// Log out the user from the application
+    fileprivate func logoutUser() {
+        
+        let navigationController = MainNavigationController()
+        
+        guard let environment = StorageProvider.storedEnvironment, let sessionToken = StorageProvider.storedSessionToken else {
+            
```

Comment:
Logging out will present a new `MainNavigationController` modally which in turn will push an `EnvironmentViewController` to the stack. If the user proceeds to select env and then log in, another `MainNavigationController` will be presented modally by the parent view controller which in this case is the original `MainNavigationController`...


Repeating this procedure will allow a user to create an ever increasing stack of modally presented `MainNavigationController`s.

```
MainNavigationController -> MainNavigationController -> MainNavigationController -> MainNavigationController
```

I would suggest using push/pop for the relevant transitions, or if modal presentation is desired, pair each `present(...)` with a `dismiss(...)`

Reply:
yes, totally agree. stupid mistake. will use push & pop

Reply:
I have considered to use pop when logging out, but as we will have two ViewControllers (EnvironmentViewController & LoginViewController ) in the stack we may need to use something like this  `popToViewController(self.navigationController?.viewControllers[0], , animated: true)` to directly navigate to  `EnvironmentViewController`. 

Instead that I used  dismissing the `LoginViewController` & making the `AssetListTableViewController` the rootView when logging the user. Let me know if you have more suggestions.